### DISCUSSION TITLE
A way to exclude a custom field from onDuplicateKeyUpdate updates

### DISCUSSION THREAD

**Discussion by GreyTeardrop** [komapper/komapper#1396] [Q&A] [answered]
Hey Komapper team! I've just recently discovered Komapper and I really like it.

At the moment I'm looking into the best option to implement fields like `createdBy`/`updatedBy` in my entity. It's expected to work similar to `@KomapperCreatedAt`/`@KomapperUpdatedAt` - the `updatedBy` field should always be updated and `createdBy` should only be touched when saving new entities.

I was able to get that to work with `onDuplicateKeyUpdate()` by explicitly listing all fields that have to be updated in the `set` block, however, I wonder if there's perhaps a simpler approach that doesn't require listing all properties that have to be set - perhaps I can either annotate the `createdBy` property with something to exclude it or the `set` block supports a way of excluding properties instead of including them?

**Comment by nakamura-to** [comment, marked as answer]
Thank you for your interest in Komapper.

You can achieve this by using the reflection and the metamodel data provided by Komapper.

In the following code example, only properties annotated with the `MyIgnore` annotation are excluded from being updated.

```kotlin
annotation class MyIgnore
```

```kotlin
@KomapperEntity
data class Employee(
    @KomapperId
    val id: Int,
    val name: String,
    @KomapperVersion
    val version: Int = 0,
    @MyIgnore
    @KomapperCreatedAt
    val createdAt: LocalDateTime = LocalDateTime.MIN,
    @KomapperUpdatedAt
    val updatedAt: LocalDateTime = LocalDateTime.MIN,
)
```


```kotlin
val e = Meta.employee

val employee = Employee(id = 1, name = "AAA")

val query = QueryDsl.insert(e).onDuplicateKeyUpdate().set {
    val paramMap = Employee::class.primaryConstructor?.parameters?.associateBy { it.name } ?: emptyMap()

    for (p in e.properties()) {
        p as PropertyMetamodel<Employee, Any, Any>
        val param = paramMap[p.name]
        if (param != null && !param.hasAnnotation<MyIgnore>()) {
            p.eq(p.getter(employee))
        }
    }
}.single(employee)

database.runQuery(query)
```

Please note, however, that since Komapper's annotations are not included in the binary, you cannot check for the presence of Komapper annotations at runtime.

**Reply by GreyTeardrop** [reply]
Thank you! I was able to avoid reflection by directly referencing properties to exclude, although it looks a bit hacky
```kotlin
val query = QueryDsl.insert(Meta.myEntity)
    .onDuplicateKeyUpdate()
    .set { target ->
        Meta.myEntity.properties().zip(target.properties()).forEach { (p, q) ->
            if (p != Meta.myEntity.createdDate && p != Meta.myEntity.createdBy) {
                @Suppress("UNCHECKED_CAST")
                (p as PropertyMetamodel<PromotionDraft, String, String>).eq(q as PropertyMetamodel<PromotionDraft, String, String>)
            }
        }
    }
    .executeAndGet(promotionDraft)
```

**Reply by GreyTeardrop** [reply]
Would you be interested in a contribution that would add a flag (`excludeFromUpdate`?) to property metamodel and add annotation (`@KomapperExcludeFromUpdate`) to populate it?

**Reply by nakamura-to** [reply]
How about adding `insertable` and `updatable` properties to `@KomapperColumn`?
When those properties are set to `false`, they will be excluded from the target of the INSERT or UPDATE statements.
By default, both would be set to `true`.

**Reply by GreyTeardrop** [reply]
That sounds better. I'll try to draft something like that.

**Reply by GreyTeardrop** [reply]
So far I've only added `updatable` attribute (https://github.com/komapper/komapper/pull/1399), no docs or tests there yet. I wonder, however, if I'm moving in the right direction with this change.

**Reply by nakamura-to** [reply]
Thank you for sharing the pull request. I think you're heading in the right direction. Keep up the good work!
